### ISSUE TITLE
Kubectl GET HPA CPU Current isn't refreshed properly under Non Successfully POD creation scenario

### ISSUE THREAD

**Issue by pepverd** [kubernetes/kubernetes#19144]
Testing Kubernetes in GKE, I think I encountered an issue related with the behaviour of the HPA in Kubectl, and particularly with the data reported by Kubectl GET HPA CPU Current that isn't refreshed properly under Non Successfully POD creation scenario. 

I was doing some load testing against one of my GKE Instances (1 simple, n1-standard-1 (1 vCPU, 3.75 GB memory)), GKE Kubectl version 1.1.2.

Under this environment, Kube-system namespace consumed about 770m during the Resource allocation (CPU Requests/Limits) running fluentd / heapster / dns / kube-ui / kubedash / l7-lb, being the rest available for the default namespace (that contains my own pods). 

I had 4 pods with some DefaultRequest limits of 50m, and initially I had configured a HPA to scale one of my pods up to 3, when the CPU Utilization reaches the target of 60%. 

Then, I launched the load testing against the system. The result was that after a couple of minutes, the system triggered new replicas but none of them was successfully created. 

Under this scenario, not only the new two replicas were not successfully created but also the `kubectl get hpa` was reporting wrong information (after the load testing was finished and the load reduced, the CPU didn't update to its actual value and the new replicas were not terminated). 

I verified that heapster was indeed reporting the right data (/api/v1/model/metrics/cpu-usage and so). I would have been expecting the pods to be cleared, and the HPA Current being refreshed according to the new data.

**Comment by fgrzadkowski** [comment]
@kubernetes/goog-control-plane

**Comment by qristin** [comment]
I'm having the same kind of problems. I have a cluster on google compute engine, which is also scaled when the CPU exceeds a certain value, just like the HPA should do. Compute engine is acting as expected, but the CPU utilization of the pods doesn't  react on increased load on the system. 
What actually happens is that the pods are restarting, which is really not what I want to happen. This is caused by a timeout on the health check. Increasing the timeout value of the livenessprobe helps, but is also not what I want. 

Does anybody know how to tweak the settings in such a way that auto scaling works properly? Other suggestion about how to tackle this problem are also welcome :)

**Comment by piosz** [comment]
I'm not sure what the problem here is. @pepverd @qristin could you please provide more details how to reproduce?

**Comment by qristin** [comment]
The issue is that the VMs are scaling much faster than the pods are scaling. But that was due to the fact that my pods were restarting. I am still not quiet sure why this is happening, but I assume it is due to a high error rate that my loadtest was generating. (How is restarting handled? And can this also be tweaked?)
Still, I would expect that the system will scale up, the VMs were doing it, so why not the pods?

**Comment by mwielgus** [comment]
Do you have both HPA and cluster autoscaling on?

**Comment by qristin** [comment]
Yes, cluster autoscaling is done on an instance group in compute engine.

**Comment by mwielgus** [comment]
In K8s < 1.2 we have an older version of Heapster that due to its suboptimal internal design takes quite a lot of time to scrape, aggregate and make available the cpu usage. With 1.2 (and Heapster 0.20.0) we have refactored/rewritten most of the code and now HPA should react significantly faster.  

Because we tried to be rather on the safe side HPA requires all metrics of available pods to be present before making any scaling decision. We are relaxing these constraints a bit.

**Comment by qristin** [comment]
Sounds good, looking forward to use it.
I still doing some testing with load on the system and it gets more clear that my pods are restarting a lot if my application throws a lot of errors (I expect these errors and rather not want to let the pods restart). I am not sure if this will also be fixed with version 1.2. Is this a known issue?

**Comment by mwielgus** [comment]
Can you tell us more about the errors? If pods are restarting like crazy we don't recommend HPA use as the CPU load will not reflect the incoming traffic but rather startup/init work.

**Comment by qristin** [comment]
The error was a timeout which was caused by another module which was not yet ready (it was not scaling fast enough). I am still not sure if a timeout error should restart an instance. On the one hand it is a valid error but restarting the instance does not help in this case. I assume that an instance is restarted if too much errors are detected. And distinguish between errors might be too difficult. Although I'm curious what the restart policy is and if it can be tweaked.
Nevertheless, a more nuanced loadtest showed the expected results and the pods weren't restarting anymore :smiley:

**Comment by arjenvanderende** [comment]
I have the same issue on GKE running master on 1.2.5 and nodes on 1.2.4. 
The current CPU% is never updated for 2 HPAs. The events of HPA 1 are empty, but the other has some events:

```
 FirstSeen	LastSeen	Count	From				SubobjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  15h		13s		1808	{horizontal-pod-autoscaler }			Warning		FailedGetMetrics	failed to get CPU consumption and request: metrics obtained for 19/25 of pods
  15h		13s		1808	{horizontal-pod-autoscaler }			Warning		FailedComputeReplicas	failed to get CPU utilization: failed to get CPU consumption and request: metrics obtained for 19/25 of pods
```

Looking at the logs of pods in the `kube-system` namespace the following stands out:

```
13:19:05.000 I0715 11:19:05.000171 1 manager.go:79] Scraping metrics start: 2016-07-15 11:18:00 +0000 UTC, end: 2016-07-15 11:19:00 +0000 UTC
13:19:05.000 E0715 11:19:05.000300 1 summary.go:390] Node gke-n1-standard-4-02c8cf7a-vqrw is not ready
13:19:05.000 E0715 11:19:05.000315 1 summary.go:390] Node gke-n1-standard-4-02c8cf7a-wmdb is not ready
13:19:05.000 E0715 11:19:05.019231 1 summary.go:114] error while getting metrics summary from Kubelet: request failed - "500 Internal Server Error", response: "Internal Error: unable to find data for container /a8ee180f06d3323ea4408137ca6f0151448a8ca9f18fc10a641b69f29850b318"
13:19:05.000 E0715 11:19:05.034353 1 summary.go:114] error while getting metrics summary from Kubelet: request failed - "500 Internal Server Error", response: "Internal Error: unable to find data for container /f526c5154d8c11767e2896a19cc1b75badb2ad7f17cb8b6b3888425701dc2bca"
13:19:05.000 I0715 11:19:05.086728 1 manager.go:152] ScrapeMetrics: time: 86.40162ms size: 173
```

Deleting and recreating the HPAs has no effect. After recreation, the current CPU% hangs in the `<waiting>` state.

**Comment by jszczepkowski** [comment]
@arjenvanderende 
HPA is failing because some pod metrics are missing. Are you sure that all 25 pods are running? In future releases #29125 will be helpful.

**Comment by DanielHeckrath** [comment]
I'm observing a similar problem on GKE running Kubernetes 1.3.5.

```
Name:				gosu-main-chat-v1
Namespace:			default
Labels:				<none>
Annotations:			<none>
CreationTimestamp:		Mon, 22 Aug 2016 12:56:25 -0700
Reference:			Deployment/gosu-main-chat-v1
Target CPU utilization:		70%
Current CPU utilization:	<unset>
Min replicas:			8
Max replicas:			20
Events:
  FirstSeen	LastSeen	Count	From				SubobjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  15m		28s		31	{horizontal-pod-autoscaler }			Warning		FailedGetMetrics	failed to get CPU consumption and request: metrics obtained for 15/16 of pods
  15m		28s		31	{horizontal-pod-autoscaler }			Warning		FailedComputeReplicas	failed to get CPU utilization: failed to get CPU consumption and request: metrics obtained for 15/16 of pods
```

The problem here seems to be that the HPA is trying to get metrics from a pod that does not exists, because the bound deployment has a target replica count of 15 

```
Name:			gosu-main-chat-v1
Namespace:		default
CreationTimestamp:	Mon, 06 Jun 2016 17:56:43 -0700
Labels:			app=gosu-main-chat
			version=v1
Selector:		app=gosu-main-chat,version=v1
Replicas:		15 updated | 15 total | 15 available | 0 unavailable
StrategyType:		RollingUpdate
MinReadySeconds:	0
RollingUpdateStrategy:	0 max unavailable, 1 max surge
OldReplicaSets:		<none>
NewReplicaSet:		gosu-main-chat-v1-2067642807 (15/15 replicas created)
Events:
  FirstSeen	LastSeen	Count	From				SubobjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  22m		22m		1	{deployment-controller }			Normal		ScalingReplicaSet	Scaled down replica set gosu-main-chat-v1-2067642807 to 10
  17m		17m		1	{deployment-controller }			Normal		ScalingReplicaSet	Scaled up replica set gosu-main-chat-v1-2067642807 to 15
```

**Comment by jszczepkowski** [comment]
The problem of autoscaling w/o info from all pods will be solved by #30471.

**Comment by jszczepkowski** [comment]
Fixed by #33593.
